### PR TITLE
Enrich too-old connection failures on instance startup

### PR DESCRIPTION
**Pull request by wethreetrees**
Part of #64567

An already-joined instance that became too old for the cluster kept attempting to connect and failing with an opaque EOF and no indication of the version mismatch.

This supplements that opaque failure with version guidance (when the instance is too old) so it's diagnosable, while leaving a rejection decision to the server and staying retryable. The pre-existing too-new check is kept, restructured within the connect path. A failed version fetch fails open so transient proxy errors can't block an otherwise healthy instance from connecting.

The too-new version check for proxy-forwarded connections now properly compares against the **Proxy Service's** version before dialing (_agent ≤ Proxy_), and against the **Auth Service's** version after connecting (_agent ≤ Auth_), per the [Upgrading Compatibility](https://goteleport.com/docs/upgrading/overview/#component-compatibility) docs:

> Other Teleport agents (SSH Service, Database Service, etc.) should always run a version less than or equal to the Proxy Service.

>Proxy Service instances and agents do not support Auth Service instances that are on an older major version, and will fail to connect to older Auth Service instances by default. For example, an 18.x.x Proxy Service or agent is not compatible with an 17.x.x Auth Service.

> [!NOTE]
> The coverage is asymmetric by design. The too-old check compares against the `MinClientVersion` which is only available from the proxy's `/webapi/find` endpoint, so this check can only apply to proxy-tunneled connections.

Changelog: Teleport instances connecting through a proxy that are running a version below the cluster's advertised minimum client version now log a clear version-compatibility error on startup instead of an opaque connection error.

### Test Environment
Local build. Three `teleport` binaries built at different major versions to induce skew (v19, v20, and v21). For most cases the instance joins once at a compatible version, then restarts against a cluster/version that puts it out of range. The proxy-lag case (instance newer than the proxy) uses a split auth/proxy deployment: a v20 auth with a v19 proxy, and a v20 instance.

### Test Cases
- [x] A compatible instance connects normally
- [x] Instance one major newer than auth is rejected on startup with a clear "too new" message (direct-to-auth)
- [x] Instance one major newer than auth is rejected on startup with a clear "too new" message (via-proxy)
- [x] A too-new rejected instance stops retrying and exits, rather than looping on reconnect
- [x] A too-old instance is not rejected; logs a clear error (with the too-old message) and retries
- [x] `--skip-version-check` lets a too-new instance connect
- [x] Instance newer than the proxy but not newer than auth is rejected on startup (proxy one major behind auth; too-new-vs-proxy check)

### CODE REVIEW

File: lib/service/connect.go [hunk 1]
```diff
@@ -1565,6 +1532,28 @@ func (process *TeleportProcess) breakerConfigForRole(role types.SystemRole) brea
...
+		process.logger.WarnContext(process.ExitContext(),
+			"Could not fetch the cluster's minimum client version from the proxy's web API, skipping the instance-too-old check for this connection attempt.",
+			"proxy_addr", proxyWebAddr.String(),
+			"error", err,
+		)
+		return nil
+	}
+	return trace.Wrap(process.enforceVersionPolicy(process.ExitContext(), joinclient.VersionInfo{
+		MinClientVersion: findResp.MinClientVersion,
+	}))
```
[the remark this answers is not in the document]
Comment (wethreetrees) [reply lib/service/connect.go]: Good catch. I missed that clause in the compatibility docs. I'll get this updated.

[the remark this answers is not in the document]
Comment (wethreetrees) [reply lib/service/connect.go]: Fixed, added the too-new check on the proxy-forwarded path.

File: lib/service/connect.go [hunk 2]
```diff
@@ -1509,6 +1471,10 @@ func (process *TeleportProcess) newClient(connector *Connector) (*authclient.Cli
 		logger.DebugContext(process.ExitContext(), "Attempting to discover reverse tunnel address.")
 		logger.DebugContext(process.ExitContext(), "Attempting to connect to Auth Server through tunnel.")
 
+		if err := process.enforceProxyVersionPolicy(authServers[0]); err != nil {
```
Comment (cthach) [inline lib/service/connect.go]: Do we need to test against the other `authServers`? Also, is this guaranteed to be non-empty? 

If this is truly safe and won't ever panic, please add a comment explaining so.

File: lib/service/connect_test.go [hunk 3]
```diff
@@ -90,77 +95,218 @@ func TestTeleportProcessClientVersionCheck(t *testing.T) {
...
+		},
+		{
+			name:          "compatible version connects",
+			serverVersion: teleport.Version,
+			assertError:   require.NoError,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			lis, err := net.Listen("tcp", "127.0.0.1:0")
+			require.NoError(t, err)
```
Comment (cthach) [inline lib/service/connect_test.go:240]:
```suggestion
			lis, err := net.Listen("tcp", "127.0.0.1:0")
			require.NoError(t, err)
			t.Cleanup(func() { _ = lis.Close() })
```

Just to be proactive about cleaning up resources on the host

File: lib/service/connect.go [hunk 4]
```diff
@@ -129,20 +127,6 @@ func (process *TeleportProcess) reconnectToAuthService(role types.SystemRole) (*
...
 }
 
-type invalidVersionErr struct {
-	ClusterMajorVersion int64
-	LocalMajorVersion   int64
-}
-
-func (i invalidVersionErr) Error() string {
-	return fmt.Sprintf("Teleport instance is too new. This instance is running v%d. The auth server is running v%d and only supports instances on v%d or v%d. To connect anyway pass the --skip-version-check flag.", i.LocalMajorVersion, i.ClusterMajorVersion, i.ClusterMajorVersion, i.ClusterMajorVersion-1)
-}
```
Comment (cthach) [inline lib/service/connect.go:140]: Can you please confirm a user will still gets an actionable error like the one this is being replaced with?  Just want to ensure we maintain that UX.

Comment (wethreetrees) [reply lib/service/connect.go:140]: Confirmed, the [too-new message](https://github.com/gravitational/teleport/blob/bfdb772ffb9ba76f81906ecd8f52f2be263c7c70/lib/service/versioncompat.go#L49-L51) is nearly identical. The only difference is "auth server" -> "server", since the check can now run against the proxy as well. The error is still fatal (not retried), so the UX holds.

> Teleport instance is too new. This instance is running v18. The server is running v17 and only supports instances on v17 or v16. To connect anyway pass the --skip-version-check flag.

File: lib/service/connect_test.go (hunk 3 above)
Comment (wethreetrees) [reply lib/service/connect_test.go:240]: I think `t.Cleanup(grpcServer.Stop)` already covers this. `Stop()` closes the listeners it was serving, per [the docs](https://pkg.go.dev/google.golang.org/grpc@v1.81.1#Server.Stop):

> Stop stops the gRPC server. It immediately closes all open connections and listeners. It cancels all active RPCs on the server side and the corresponding pending RPCs on the client side will get notified by connection errors.

I added an explicit `lis.Close()` anyway, it's harmless and makes the intent clear.

File: lib/service/connect.go (hunk 2 above)
Comment (wethreetrees) [reply lib/service/connect.go]: Good questions. I added a [comment](https://github.com/gravitational/teleport/blob/bfdb772ffb9ba76f81906ecd8f52f2be263c7c70/lib/service/connect.go#L1474-L1476) that should cover them all:

```
// authServers has at least one entry (config validation requires auth_servers
// for v1/v2), and the reverse tunnel webclient resolver dials authServers[0],
// so that's the endpoint whose policy applies to the tunnel we're opening.
```

If we'd rather not index here at all, we could stash the address on the process struct when it's first derived and reference that instead.

File: lib/service/connect_test.go (hunk 3 above)
Comment (cthach) [reply lib/service/connect_test.go:240]: That makes sense. I was more so proposing this as a defensive just in case something happened during `grpcServer.Serve(lis)` that would cause `t.Cleanup(grpcServer.Stop)` to not fire properly to clean up the underlying listener. I'm paranoid about the failure cases 🤪

File: lib/service/connect.go (hunk 4 above)
Comment (cthach) [reply lib/service/connect.go:140]: Thank you!

**Review by cthach — APPROVED** [review]
Well done!

I don't know how often v1/v2 configs appear in the real world but consider adding test coverage for the fall back path to ensure the `enforceProxyVersionPolicy()` is actually applied in `newClient()`.

File: lib/service/connect.go (hunk 2 above)
Comment (cthach) [reply lib/service/connect.go]: Comment is good for me. Thanks for looking into it!

**Review by espadolini — COMMENTED** [review]
IIRC the choice of only checking for version compatibility in one way on each side of the connection was deliberate, so one does not need to go edit the configuration on a lot of clients that are being held back from upgrading to allow them to connect, only the auths in the control plane have to be configured with the unstable envvar. This seems to break that and after an upgrade (maybe to the wrong major version, but to a version containing this change) agents that were perfectly happy to connect to a control plane that was not rejecting them will hard fail and crashloop. Are we sure we want that to happen?

**Comment by wethreetrees** [comment]
> IIRC the choice of only checking for version compatibility in one way on each side of the connection was deliberate, so one does not need to go edit the configuration on a lot of clients that are being held back from upgrading to allow them to connect, only the auths in the control plane have to be configured with the unstable envvar. This seems to break that and after an upgrade (maybe to the wrong major version, but to a version containing this change) agents that were perfectly happy to connect to a control plane that was not rejecting them will hard fail and crashloop. Are we sure we want that to happen?

I've implemented what we discussed in our team meeting last week. The client-side too-old check is now advisory-only: when the server force closes connections to a too-old agent, we supplement the errors with version guidance and the agent retries instead of crashlooping.

The changes are significant enough that this requires a re-review; description and test plan updated above.

PTAL @cthach @espadolini 🙏

File: lib/service/connect.go [hunk 5]
```diff
@@ -1611,7 +1597,14 @@ func (process *TeleportProcess) newClientThroughProxy(tlsConfig *tls.Config, ssh
...
 	resp, err := clt.Ping(ctx)
 	if err != nil {
-		return nil, nil, trace.NewAggregate(err, clt.Close())
+		return nil, nil, trace.NewAggregate(
+			err,
+			clt.Close(),
+			// A too-old instance is force-closed by the server with an opaque EOF.
+			// Include a too-old error (nil if not too old) so an operator knows
+			// that version compatibility is a potential cause of the failure.
+			clientTooOld(versionInfo.MinClientVersion),
```
[the remark this answers is not in the document]
Comment (wethreetrees) [reply lib/service/connect.go:1613]: This is correct and I've moved the too-new check against auth back into `getConnector` so it will apply to both direct-to-auth and proxy-forwarded connections.

File: lib/service/servicecfg/config.go [hunk 6]
```diff
@@ -638,6 +635,21 @@ func (cfg *Config) SetAuthServerAddress(addr utils.NetAddr) {
 	cfg.authServers = []utils.NetAddr{addr}
 }
 
+// ProxyWebAddr returns the address of the proxy web API: the configured proxy
+// for v3 configs, otherwise the first auth server address.
```
Comment (espadolini) [inline lib/service/servicecfg/config.go]: I'd add a note about this function potentially returning the address of the auth rather than the address of the proxy.

File: lib/service/connect.go [hunk 7]
```diff
@@ -1573,7 +1533,40 @@ func (process *TeleportProcess) breakerConfigForRole(role types.SystemRole) brea
...
+		return joinclient.VersionInfo{}
+	}
+	return joinclient.VersionInfo{
+		ServerVersion:    findResp.ServerVersion,
+		MinClientVersion: findResp.MinClientVersion,
+	}
+}
+
 func (process *TeleportProcess) newClientThroughProxy(tlsConfig *tls.Config, sshConfig ssh.ClientConfig, role types.SystemRole, clusterName string, getClusterCAs func() (*x509.CertPool, error)) (*authclient.Client, *proto.PingResponse, error) {
+	versionInfo := process.proxyVersionInfo()
```
[the remark this answers is not in the document]
Comment (espadolini) [reply lib/service/connect.go:1565]: How does this change anything? The proxy reached by the resolver might be replaced by one of a different version immediately after the resolver returns.

File: lib/service/connect.go [hunk 8]
```diff
@@ -1573,7 +1533,40 @@ func (process *TeleportProcess) breakerConfigForRole(role types.SystemRole) brea
...
+		Insecure:  process.Config.InsecureMode,
+	})
+	if err != nil {
+		process.logger.WarnContext(process.ExitContext(),
+			"Could not fetch version information from the proxy's web API, skipping the version compatibility check for this connection attempt.",
+			"proxy_addr", proxyAddr.String(),
+			"error", err,
+		)
+		return joinclient.VersionInfo{}
+	}
```
Comment (espadolini) [inline lib/service/connect.go:1557]: This will unconditionally log at warn level on every agent startup for agents pointed directly at an auth, which I don't think we should do.

File: lib/service/connect.go [hunk 9]
```diff
@@ -1611,7 +1604,14 @@ func (process *TeleportProcess) newClientThroughProxy(tlsConfig *tls.Config, ssh
...
 	resp, err := clt.Ping(ctx)
 	if err != nil {
-		return nil, nil, trace.NewAggregate(err, clt.Close())
+		return nil, nil, trace.NewAggregate(
+			err,
+			clt.Close(),
+			// A too-old instance is force-closed by the server with an opaque EOF.
+			// Include a too-old error (nil if not too old) so an operator knows
+			// that version compatibility is a potential cause of the failure.
+			clientTooOld(versionInfo.MinClientVersion),
```
Comment (espadolini) [inline lib/service/connect.go:1613]: Should we only include the version mismatch error on an EOF?

File: lib/service/connect.go (hunk 8 above)
Comment (wethreetrees) [reply lib/service/connect.go:1557]: This only runs in `newClientThroughProxy` and in the v1/v2 path, we unconditionally attempt a direct-to-auth connection via `newClientDirect` before trying through the proxy.

File: lib/service/servicecfg/config.go (hunk 6 above)
Comment (wethreetrees) [reply lib/service/servicecfg/config.go]: I rewrote the first section of the comment for clarity https://github.com/gravitational/teleport/pull/68461/commits/3180706751c632980063972c489dccb9df4f13b4

File: lib/service/connect.go (hunk 9 above)
Comment (wethreetrees) [reply lib/service/connect.go:1613]: Yeah, I think it would be good to narrow it. Looks like string matching for `EOF` in the error message is our best bet (`errors.Is(io.EOF)` doesn't work since grpc strips the `io.EOF` sentinel when it returns). In combination with `IsConnectionProblem`, I think this narrows the version error supplementation to the smallest number of scenarios that include the `ValidateClientVersion` force-close.

Let me know if this sounds good or if you had something else in mind.

Comment (wethreetrees) [reply lib/service/connect.go:1613]: I spoke with @rosstimothy today and he voted for leaving it open. Since it's advisory only, it probably can't hurt to just add the version info with any error we get.

I'm leaving this as-is for now, but happy to make a change if we want to keep the version info closer to the `ValidateClientVersion` force-close errors.

**Review by cthach — APPROVED** [review]

**Review by espadolini — APPROVED** [review]
